src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -186,6 +214,25 @@ private async Task ProcessQueueAsync()
                     var context = await work.CreateRequestContextAsync(cancellationToken).ConfigureAwait(false);
                     if (work.MutatesServerState)
                     {
+                        if (CancelInProgressWorkUponMutatingRequest)
+                        {
+                            // Cancel all concurrently executing tasks
+                            var concurrentlyExecutingTasksArray = concurrentlyExecutingTasks.ToArray();
+                            for (var i = 0; i < concurrentlyExecutingTasksArray.Length; i++)
+                            {
+                                concurrentlyExecutingTasksArray[i].Value.Cancel();
+                            }
+
+                            try
+                            {
+                                // wait for all pending tasks to complete their cancellation
+                                await Task.WhenAll(concurrentlyExecutingTasksArray.Select(kvp => kvp.Key)).ConfigureAwait(false);
+                            }
+                            catch (TaskCanceledException)
```

Comment:
I think we have a no-throw await helper somewhere to avoid the catch?

Reply:
Ooh, can you point me to it? Keep in mind that this assembly has minimal roslyn dependencies.

Reply:
https://github.com/dotnet/roslyn/blob/e9f1cd2980203eb9f7d9da81ed898f6f28a29b1f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/TestHooks/TaskExtensions.cs#L26 I think?

Reply:
Thanks, that works wonderfully